CrossCompatibility/CrossCompatibility/Common/PowerShellVersion.cs

```diff
@@ -0,0 +1,369 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Text;
+using Newtonsoft.Json;
+
+namespace Microsoft.PowerShell.CrossCompatibility
+{
+    /// <summary>
+    /// Class created to bridge the gap between System.Version and System.Management.Automation.SemanticVersion.
+    /// Allows a version encoding either to be deserialized into this type.
+    /// </summary>
+    public class PowerShellVersion
+    {
+        /// <summary>
+        /// Create a PowerShellVersion from another version object.
+        /// Currently accepts a dynamic input to allow SemanticVersions.
+        /// </summary>
+        /// <param name="versionInput">A version-like object describing a PowerShell version.</param>
+        /// <returns>A PowerShellVersion object.</returns>
+        public static PowerShellVersion Create(dynamic versionInput)
+        {
+            switch (versionInput)
+            {
+                case Version systemVersion:
+                    return (PowerShellVersion)systemVersion;
+
+                case string versionString:
+                    return Parse(versionString);
+            }
+
+            if (versionInput.BuildLabel != null)
+            {
+                return new PowerShellVersion(versionInput.Major, versionInput.Minor, versionInput.Patch, $"{versionInput.PreReleaseLabel}+{versionInput.BuildLabel}");
+            }
+
+            return new PowerShellVersion(versionInput.Major, versionInput.Minor, versionInput.Patch, versionInput.PreReleaseLabel);
+        }
+
+        /// <summary>
+        /// Parse a PowerShellVersion from a string.
+        /// </summary>
+        /// <param name="versionStr">The version-describing string to parse.</param>
+        /// <returns>A PowerShellVersion, as described by the string.</returns>
+        public static PowerShellVersion Parse(string versionStr)
+        {
+            if (versionStr == null)
+            {
+                throw new ArgumentNullException(nameof(versionStr));
+            }
+
+            int[] versionParts = new int[3] { -1, -1, -1 };
```

Comment:
Agreed this would be simpler. I'm going to leave it as is for now since it's currently working in its present form with some tests. We can change it later if required I think.